README
<<<<<<< HEAD
h1. Welcome to Flume!

Flume is a distributed, reliable, and available service for
efficiently collecting, aggregating, and moving large amounts of log
data. It has a simple and flexible architecture based on streaming
data flows. It is robust and fault tolerant with tunable reliability
mechanisms and many failover and recovery mechanisms. The system is
centrally managed and allows for intelligent dynamic management. It
uses a simple extensible data model that allows for online analytic
application

Flume is open-sourced under the Apache Software Foundation License v2.0.  

h2.  Documentation

Flume's installation instructions can be found here:

https://docs.cloudera.com/display/DOC/Flume+Installation

Flume's user guide, cookbook, and FAQ are available here:

* http://archive.cloudera.com/cdh/3/flume/UserGuide/index.html
* http://archive.cloudera.com/cdh/3/flume/Cookbook/index.html
* http://github.com/cloudera/flume/wiki/FAQ

h2. Contact us!

* User mailing list: https://groups.google.com/a/cloudera.org/group/flume-user/topics
* Developer mailling list: https://groups.google.com/a/cloudera.org/group/flume-dev/topics
=======
h1. Welcome to Apache Flume!

Apache Flume is a distributed, reliable, and available service for efficiently
collecting, aggregating, and moving large amounts of log data. It has a simple
and flexible architecture based on streaming data flows. It is robust and fault
tolerant with tunable reliability mechanisms and many failover and recovery
mechanisms. The system is centrally managed and allows for intelligent dynamic
management. It uses a simple extensible data model that allows for online
analytic application.

The Apache Flume 1.x (NG) code line is a refactoring of the first generation
Flume to solve certain known issues and limitations of the original design.

Apache Flume is open-sourced under the Apache Software Foundation License v2.0.

h2. Documentation

Documentation is included in the binary distribution under the docs directory.
In source form, it can be found in the flume-ng-doc directory.

The Flume 1.x guide and FAQ are available here:

* https://cwiki.apache.org/FLUME/flume-ng.html
* https://cwiki.apache.org/confluence/display/FLUME/Getting+Started

h2. Contact us!

* Mailing lists: https://cwiki.apache.org/confluence/display/FLUME/Mailing+Lists
>>>>>>> refs/remotes/apache/trunk
* IRC channel #flume on irc.freenode.net

Bug and Issue tracker.

<<<<<<< HEAD
* http://issues.cloudera.org/
* https://issues.cloudera.org/browse/FLUME
=======
* https://issues.apache.org/jira/browse/FLUME
>>>>>>> refs/remotes/apache/trunk

h2. Compiling Flume

Compiling Flume requires the following tools:

<<<<<<< HEAD
* Java JDK 1.6
* Apache Maven

To compile Flume, run `mvn compile`.
To build a distribution, run `mvn package`.

The final Flume distribution artifact will be in $project/flume-distribution/target/.
=======
* Oracle Java JDK 1.6
* Apache Maven 3.x

Note: The Apache Flume build requires more memory than the default configuration.
We recommend you set the following Maven options:

export MAVEN_OPTS="-Xms512m -Xmx1024m -XX:PermSize=256m -XX:MaxPermSize=512m"

To compile Flume, run `mvn compile`.
To build a distribution, run `mvn install`.

The final Flume distribution artifacts will be in $project/flume-ng-dist/target/.
>>>>>>> refs/remotes/apache/trunk


CHANGELOG
<<<<<<< HEAD
Release Notes - Flume - Version 1.1.0-incubating
=======
Release Notes - Flume - Version v1.6.0

** Sub-task
    * [FLUME-2250] - Add support for Kafka Source
    * [FLUME-2251] - Add support for Kafka Sink
    * [FLUME-2677] - Update versions in 1.6.0 branch
    * [FLUME-2686] - Update KEYS file for 1.6 release

** Bug
    * [FLUME-1793] - Unit test TestElasticSearchLogStashEventSerializer fails with IBM JDK
    * [FLUME-1934] - Spoolingdir source exception when reading multiple   zero size files
    * [FLUME-2024] - Add ExecSource flush timeout to Flume User Guide
    * [FLUME-2126] - Problem in elasticsearch sink when the event body is a complex field
    * [FLUME-2148] - Windows : Add  flume-env.ps1
    * [FLUME-2214] - FileChannelIntegrityTool needs documentation in user guide
    * [FLUME-2245] - HDFS files with errors unable to close
    * [FLUME-2337] - export JAVA_HOME in flume-env.sh.template and increase heap size
    * [FLUME-2346] - idLogFileMap in Log can lose track of file ids
    * [FLUME-2404] - Default maxReadBufferBytes might cause OOM and cause scribe source exit
    * [FLUME-2408] - Remove FLUME-1899 from Flume 1.5.0 change log.
    * [FLUME-2416] - Use CodecPool in compressed stream to prevent leak of direct buffers
    * [FLUME-2420] - HDFSEventSink#stop might throw ConcurrentModificationException
    * [FLUME-2425] - FileChannel should trim data and checkpoint directories
    * [FLUME-2432] - Update Kite dependency to 0.15.0
    * [FLUME-2438] - Make Syslog source message body configurable
    * [FLUME-2450] - Improve replay index insertion speed.
    * [FLUME-2466] - Embedded agent name cannot contain spaces
    * [FLUME-2470] - Kafka Sink and Source must use camel case for all configs.
    * [FLUME-2472] - SyslogUtils fails when system locale is not English
    * [FLUME-2475] - toLowerCase/toUpperCase used without Locale parameter to manipulate Enum
    * [FLUME-2479] - Kafka property auto.commit.enable is incorrect for KafkaSource
    * [FLUME-2480] - Typo in file channel exception: "Usable space exhaused"
    * [FLUME-2481] - TestFileChannelEncryption fails with System locale other than en_US
    * [FLUME-2482] - Race condition in File Channels' Log.removeOldLogs
    * [FLUME-2486] - TestExecSource fails on some environments
    * [FLUME-2487] - SyslogParser has rounding errors in timestamp parsing
    * [FLUME-2488] - TestElasticSearchRestClient fails on Oracle JDK 8
    * [FLUME-2489] - Upgrade Apache Curator to 2.6.0
    * [FLUME-2492] - Flume's Kafka Source doesn't account time correctly
    * [FLUME-2495] - Kafka Source may miss events when channel is not available
    * [FLUME-2497] - TCP and UDP syslog sources parsing the timestamp incorrectly
    * [FLUME-2500] - Add a channel that uses Kafka
    * [FLUME-2505] - Test added in FLUME-2502 is flaky
    * [FLUME-2517] - Performance issue: SimpleDateFormat constructor takes 30% of HDFSEventSink.process()
    * [FLUME-2525] - flume should handle a zero byte .flumespool-main.meta file for the spooldir source
    * [FLUME-2538] - TestResettableFileInputStream fails on JDK 8
    * [FLUME-2556] - TestBucketPath.testDateFormatHours fails intermittently
    * [FLUME-2557] - DatasetSink doesn't work with Kerberos
    * [FLUME-2560] - Kafka channel entries missing for output jar generation
    * [FLUME-2568] - Additional fix for TestReliableSpoolingFileEventReader
    * [FLUME-2578] - Kafka source throws NPE if Kafka record has null key
    * [FLUME-2586] - HDFS Sink should have an option to try rename even if close fails
    * [FLUME-2592] - Specify main manifest attribute in flume tools jar
    * [FLUME-2594] - Close Async HBase Client if there are large number of consecutive timeouts
    * [FLUME-2595] - Add option to checkpoint on file channel shutdown
    * [FLUME-2624] - Improve Hive Sink performance
    * [FLUME-2626] - Remove trustmanager-type from Thrift RPC client
    * [FLUME-2633] - Update Kite dependency to 1.0.0
    * [FLUME-2639] - SecureThriftRpcClient client privileged calls throw IllegalAccessError
    * [FLUME-2645] - ipFilter.rules property name is wrong
    * [FLUME-2654] - Memory channel error message related to byteCapacity needs refining
    * [FLUME-2657] - Upgrade to Hive 1.0
    * [FLUME-2658] - Upgrade to Hbase 1.0
    * [FLUME-2664] - Site profile fails due to dependency issues
    * [FLUME-2665] - Update documentation for hdfs.closeTries based on FLUME-2586
    * [FLUME-2670] - Modify dependencies to not pull hadoop/hive related jars
    * [FLUME-2673] - Remove unused import in TestLog
    * [FLUME-2678] - Upgrade xalan to 2.7.2 to take care of CVE-2014-0107 vulnerability
    * [FLUME-2679] - Make hbase-1 the default profile

** Dependency upgrade
    * [FLUME-2443] - org.apache.hadoop.fs.FSDataOutputStream.sync() is deprecated in hadoop 2.4

** Documentation
    * [FLUME-1594] - Document JMX options
    * [FLUME-2509] - Add description for properties in Failover Sink Processor
    * [FLUME-2630] - Update documentation for Thrift Src/Sink SSL support and Kerberos support
    * [FLUME-2635] - Documentation for zookeeper dynamic configuration in flume
    * [FLUME-2668] - Document SecureThriftRpcClient/SecureRpcClientFactory in Flume Developer Guide

** Improvement
    * [FLUME-1334] - Write an startscript for flume agents on Windows
    * [FLUME-1491] - Dynamic configuration from Zookeeper watcher
    * [FLUME-1521] - Document the StressSource
    * [FLUME-1710] - JSONEvent.getBody should not return null
    * [FLUME-1920] - Test case TestFileChannel fails when flume is built from paths containing the string "hadoop"
    * [FLUME-2237] - Example config in User Guide to use same notation for Agent name
    * [FLUME-2246] - event body data size can make it configurable for logger sinker
    * [FLUME-2273] - ElasticSearchSink: Add handling for header substitution in indexName

flume-ng-core/src/main/java/org/apache/flume/source/SyslogUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.flume.source;

import org.apache.flume.Event;
<<<<<<< HEAD
=======
import org.apache.flume.annotations.InterfaceAudience;
import org.apache.flume.annotations.InterfaceStability;
>>>>>>> refs/remotes/apache/trunk
import org.apache.flume.event.EventBuilder;
import org.jboss.netty.buffer.ChannelBuffer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
<<<<<<< HEAD
=======
import java.util.Set;
>>>>>>> refs/remotes/apache/trunk
import java.util.regex.MatchResult;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

<<<<<<< HEAD
=======
@InterfaceAudience.Private
@InterfaceStability.Evolving
>>>>>>> refs/remotes/apache/trunk
public class SyslogUtils {
  final public static String SYSLOG_TIMESTAMP_FORMAT_RFC5424_2 = "yyyy-MM-dd'T'HH:mm:ss.SZ";
  final public static String SYSLOG_TIMESTAMP_FORMAT_RFC5424_1 = "yyyy-MM-dd'T'HH:mm:ss.S";
  final public static String SYSLOG_TIMESTAMP_FORMAT_RFC5424_3 = "yyyy-MM-dd'T'HH:mm:ssZ";
  final public static String SYSLOG_TIMESTAMP_FORMAT_RFC5424_4 = "yyyy-MM-dd'T'HH:mm:ss";
  final public static String SYSLOG_TIMESTAMP_FORMAT_RFC3164_1 = "yyyyMMM d HH:mm:ss";

  final public static String SYSLOG_MSG_RFC5424_0 =
<<<<<<< HEAD
      "(?:\\d\\s)?" +// version
  // yyyy-MM-dd'T'HH:mm:ss.SZ or yyyy-MM-dd'T'HH:mm:ss.S+hh:mm or - (null stamp)
      "(?:(\\d{4}[-]\\d{2}[-]\\d{2}[T]\\d{2}[:]\\d{2}[:]\\d{2}(?:\\.\\d{1,6})?(?:[+-]\\d{2}[:]\\d{2}|Z)?)|-)" + // stamp
=======
      "(?:\\<(\\d{1,3})\\>)" + // priority
      "(?:(\\d?)\\s?)" + // version
      /* yyyy-MM-dd'T'HH:mm:ss.SZ or yyyy-MM-dd'T'HH:mm:ss.S+hh:mm or - (null stamp) */
      "(?:" +
        "(\\d{4}[-]\\d{2}[-]\\d{2}[T]\\d{2}[:]\\d{2}[:]\\d{2}" +
        "(?:\\.\\d{1,6})?(?:[+-]\\d{2}[:]\\d{2}|Z)?)|-)" + // stamp
>>>>>>> refs/remotes/apache/trunk
      "\\s" + // separator
      "(?:([\\w][\\w\\d\\.@\\-]*)|-)" + // host name or - (null)
      "\\s" + // separator
      "(.*)$"; // body

  final public static String SYSLOG_MSG_RFC3164_0 =
      "(?:\\<(\\d{1,3})\\>)" +
      "(?:(\\d)?\\s?)" + // version
      // stamp MMM d HH:mm:ss, single digit date has two spaces
      "([A-Z][a-z][a-z]\\s{1,2}\\d{1,2}\\s\\d{2}[:]\\d{2}[:]\\d{2})" +
      "\\s" + // separator
      "([\\w][\\w\\d\\.@-]*)" + // host
      "\\s(.*)$";  // body

  final public static int SYSLOG_PRIORITY_POS = 1;
  final public static int SYSLOG_VERSION_POS = 2;
  final public static int SYSLOG_TIMESTAMP_POS = 3;
  final public static int SYSLOG_HOSTNAME_POS = 4;
  final public static int SYSLOG_BODY_POS = 5;

  private Mode m = Mode.START;
  private StringBuilder prio = new StringBuilder();
  private ByteArrayOutputStream baos;
  private static final Logger logger = LoggerFactory
      .getLogger(SyslogUtils.class);

  final public static String SYSLOG_FACILITY = "Facility";
  final public static String SYSLOG_SEVERITY = "Severity";
  final public static String SYSLOG_PRIORITY = "Priority";
  final public static String SYSLOG_VERSION = "Version";
  final public static String EVENT_STATUS = "flume.syslog.status";
  final public static Integer MIN_SIZE = 10;
  final public static Integer DEFAULT_SIZE = 2500;
  private final boolean isUdp;
  private boolean isBadEvent;
  private boolean isIncompleteEvent;
  private Integer maxSize;
  private Set<String> keepFields;

  private class SyslogFormatter {
    public Pattern regexPattern;
    public ArrayList<String> searchPattern = new ArrayList<String>();
    public ArrayList<String> replacePattern = new ArrayList<String>();
    public ArrayList<SimpleDateFormat> dateFormat = new ArrayList<SimpleDateFormat>();
    public boolean addYear;
  }
  private ArrayList<SyslogFormatter> formats = new ArrayList<SyslogFormatter>();

  private String priority = null;
  private String version = null;
  private String timeStamp = null;
  private String hostName = null;
  private String msgBody = null;

  private static final String[] DEFAULT_FIELDS_TO_KEEP = {
    SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_PRIORITY,
    SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_VERSION,
    SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_TIMESTAMP,
    SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_HOSTNAME
  };
  public static final String KEEP_FIELDS_ALL = "--all--";

  public static boolean keepAllFields(Set<String> keepFields) {
    if (keepFields == null) {
      return false;
    }
    return keepFields.contains(KEEP_FIELDS_ALL);
  }

  public static Set<String> chooseFieldsToKeep(String keepFields) {
    if (keepFields == null) {
      return null;
    }

    keepFields = keepFields.trim().toLowerCase(Locale.ENGLISH);

    if (keepFields.equals("false") || keepFields.equals("none")) {
      return null;
    }

    if (keepFields.equals("true") || keepFields.equals("all")) {
      Set<String> fieldsToKeep = new HashSet<String>(1);
      fieldsToKeep.add(KEEP_FIELDS_ALL);
      return fieldsToKeep;
    }

    Set<String> fieldsToKeep = new HashSet<String>(DEFAULT_FIELDS_TO_KEEP.length);

    for (String field : DEFAULT_FIELDS_TO_KEEP) {
      if (keepFields.indexOf(field) != -1) {
        fieldsToKeep.add(field);
      }
    }

    return fieldsToKeep;
  }

  public static String addFieldsToBody(Set<String> keepFields,
                                       String body,
                                       String priority,
                                       String version,
                                       String timestamp,
                                       String hostname) {
    // Prepend fields to be kept in message body.
    if (keepFields != null) {
      if (keepFields.contains(SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_HOSTNAME)) {
        body = hostname + " " + body;
      }
      if (keepFields.contains(SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_TIMESTAMP)) {
        body = timestamp + " " + body;
      }
      if (keepFields.contains(SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_VERSION)) {
        if (version != null && !version.isEmpty()) {
          body = version + " " + body;
        }
      }
      if (keepFields.contains(SyslogSourceConfigurationConstants.CONFIG_KEEP_FIELDS_PRIORITY)) {
        body = "<" + priority + ">" + body;
      }
    }

    return body;
  }

  public SyslogUtils() {
    this(false);
  }

  public SyslogUtils(boolean isUdp) {
    this(DEFAULT_SIZE,
        new HashSet<String>(Arrays.asList(SyslogSourceConfigurationConstants.DEFAULT_KEEP_FIELDS)),
        isUdp);
  }

  public SyslogUtils(Integer eventSize, Set<String> keepFields, boolean isUdp) {
    this.isUdp = isUdp;
    isBadEvent = false;
    isIncompleteEvent = false;
    maxSize = (eventSize < MIN_SIZE) ? MIN_SIZE : eventSize;
    baos = new ByteArrayOutputStream(eventSize);
    this.keepFields = keepFields;
    initHeaderFormats();
  }

  // extend the default header formatter
  public void addFormats(Map<String, String> formatProp) {
    if (formatProp.isEmpty() || !formatProp.containsKey(
        SyslogSourceConfigurationConstants.CONFIG_REGEX)) {
      return;
    }
    SyslogFormatter fmt1 = new SyslogFormatter();
    fmt1.regexPattern = Pattern.compile( formatProp.get(
        SyslogSourceConfigurationConstants.CONFIG_REGEX) );
    if (formatProp.containsKey(
        SyslogSourceConfigurationConstants.CONFIG_SEARCH)) {
      fmt1.searchPattern.add(formatProp.get(
          SyslogSourceConfigurationConstants.CONFIG_SEARCH));
    }
    if (formatProp.containsKey(
        SyslogSourceConfigurationConstants.CONFIG_REPLACE)) {
      fmt1.replacePattern.add(formatProp.get(
          SyslogSourceConfigurationConstants.CONFIG_REPLACE));
    }
    if (formatProp.containsKey(
        SyslogSourceConfigurationConstants.CONFIG_DATEFORMAT)) {
        fmt1.dateFormat.add(new SimpleDateFormat(formatProp.get(
            SyslogSourceConfigurationConstants.CONFIG_DATEFORMAT)));
    }
    formats.add(0, fmt1);
  }

  // setup built-in formats
  private void initHeaderFormats() {
    // setup RFC5424 formater
    SyslogFormatter fmt1 = new SyslogFormatter();
    fmt1.regexPattern = Pattern.compile(SYSLOG_MSG_RFC5424_0);
    // 'Z' in timestamp indicates UTC zone, so replace it it with '+0000' for date formatting
    fmt1.searchPattern.add("Z");
    fmt1.replacePattern.add("+0000");
    // timezone in RFC5424 is [+-]tt:tt, so remove the ':' for java date formatting
    fmt1.searchPattern.add("([+-])(\\d{2})[:](\\d{2})");
    fmt1.replacePattern.add("$1$2$3");
    // FLUME-2497: SimpleDateFormat does not handle microseconds, Truncate after 3 digits.
    fmt1.searchPattern.add("(T\\d{2}:\\d{2}:\\d{2}\\.\\d{3})(\\d*)");
    fmt1.replacePattern.add("$1");
    fmt1.dateFormat.add(new SimpleDateFormat(SYSLOG_TIMESTAMP_FORMAT_RFC5424_1, Locale.ENGLISH));
    fmt1.dateFormat.add(new SimpleDateFormat(SYSLOG_TIMESTAMP_FORMAT_RFC5424_2, Locale.ENGLISH));
    fmt1.dateFormat.add(new SimpleDateFormat(SYSLOG_TIMESTAMP_FORMAT_RFC5424_3, Locale.ENGLISH));
    fmt1.dateFormat.add(new SimpleDateFormat(SYSLOG_TIMESTAMP_FORMAT_RFC5424_4, Locale.ENGLISH));
    fmt1.addYear = false;

    // setup RFC3164 formater
    SyslogFormatter fmt2 = new SyslogFormatter();
    fmt2.regexPattern = Pattern.compile(SYSLOG_MSG_RFC3164_0);
    // the single digit date has two spaces, so trim it
    fmt2.searchPattern.add("  ");
    fmt2.replacePattern.add(" ");
    fmt2.dateFormat.add(new SimpleDateFormat(SYSLOG_TIMESTAMP_FORMAT_RFC3164_1, Locale.ENGLISH));
    fmt2.addYear = true;

    formats.add(fmt1);
    formats.add(fmt2);
  }

  enum Mode {
    START, PRIO, DATA
  };

  public enum SyslogStatus{
    OTHER("Unknown"),
    INVALID("Invalid"),
    INCOMPLETE("Incomplete");

    private final String syslogStatus;

    private SyslogStatus(String status){
      syslogStatus = status;
    }

    public String getSyslogStatus(){
      return this.syslogStatus;
    }
  }

  // create the event from syslog data
  Event buildEvent() {
<<<<<<< HEAD
    byte[] body;
    int pri = 0;
    int sev = 0;
    int facility = 0;

    if(!isBadEvent){
      pri = Integer.parseInt(prio.toString());
      sev = pri % 8;
      facility = pri / 8;
      formatHeaders();
    }
=======
    try {
      byte[] body;
      int pri = 0;
      int sev = 0;
      int facility = 0;

      if(!isBadEvent){
        pri = Integer.parseInt(prio.toString());
        sev = pri % 8;
        facility = pri / 8;
        formatHeaders();
      }
>>>>>>> refs/remotes/apache/trunk

      Map <String, String> headers = new HashMap<String, String>();
      headers.put(SYSLOG_FACILITY, String.valueOf(facility));
      headers.put(SYSLOG_SEVERITY, String.valueOf(sev));
      if ((priority != null) && (priority.length() > 0)) {
        headers.put("priority", priority);
      }
      if ((version != null) && (version.length() > 0)) {
        headers.put("version", version);
      }
      if ((timeStamp != null) && timeStamp.length() > 0) {
        headers.put("timestamp", timeStamp);
      }
      if ((hostName != null) && (hostName.length() > 0)) {
        headers.put("host", hostName);
      }
      if(isBadEvent){
        logger.warn("Event created from Invalid Syslog data.");
        headers.put(EVENT_STATUS, SyslogStatus.INVALID.getSyslogStatus());
      } else if(isIncompleteEvent){
        logger.warn("Event size larger than specified event size: {}. You should " +
            "consider increasing your event size.", maxSize);
        headers.put(EVENT_STATUS, SyslogStatus.INCOMPLETE.getSyslogStatus());
      }

      if (!keepAllFields(keepFields)) {
        if ((msgBody != null) && (msgBody.length() > 0)) {
          body = msgBody.getBytes();
        } else {
          // Parse failed.
          body = baos.toByteArray();
        }
      } else {
        body = baos.toByteArray();
      }
      // format the message
      return EventBuilder.withBody(body, headers);
    } finally {
      reset();
    }
  }

  // Apply each known pattern to message
  private void formatHeaders() {
    String eventStr = baos.toString();
<<<<<<< HEAD
=======
    String timeStampString = null;

>>>>>>> refs/remotes/apache/trunk
    for(int p=0; p < formats.size(); p++) {
      SyslogFormatter fmt = formats.get(p);
      Pattern pattern = fmt.regexPattern;
      Matcher matcher = pattern.matcher(eventStr);
      if (! matcher.matches()) {
        continue;
      }
      MatchResult res = matcher.toMatchResult();
      for (int grp=1; grp <= res.groupCount(); grp++) {
        String value = res.group(grp);
        if (grp == SYSLOG_TIMESTAMP_POS) {
          timeStampString = value;

          // apply available format replacements to timestamp
          if (value != null) {
            for (int sp=0; sp < fmt.searchPattern.size(); sp++) {
              value = value.replaceAll(fmt.searchPattern.get(sp), fmt.replacePattern.get(sp));
            }
            // Add year to timestamp if needed
            if (fmt.addYear) {
              value = String.valueOf(Calendar.getInstance().get(Calendar.YEAR)) + value;
            }
            // try the available time formats to timestamp
            for (int dt = 0; dt < fmt.dateFormat.size(); dt++) {
              try {
                timeStamp = String.valueOf(fmt.dateFormat.get(dt).parse(value).getTime());
                break; // done. formatted the time
              } catch (ParseException e) {
                // Error formatting the timeStamp, try next format
                continue;
              }
            }
          }
        } else if (grp == SYSLOG_HOSTNAME_POS) {
          hostName = value;
        } else if (grp == SYSLOG_PRIORITY_POS) {
          priority = value;
        } else if (grp == SYSLOG_VERSION_POS) {
          version = value;
        } else if (grp == SYSLOG_BODY_POS) {
          msgBody = addFieldsToBody(keepFields, value, priority, version, timeStampString, hostName);
        }
      }
      break; // we successfully parsed the message using this pattern
    }
  }

  private void reset(){
    baos.reset();
    m = Mode.START;
    prio.delete(0, prio.length());
    isBadEvent = false;
    isIncompleteEvent = false;
    hostName = null;
    timeStamp = null;
    msgBody = null;
  }

  // extract relevant syslog data needed for building Flume event
  public Event extractEvent(ChannelBuffer in){

    /* for protocol debugging
    ByteBuffer bb = in.toByteBuffer();
    int remaining = bb.remaining();
    byte[] buf = new byte[remaining];
    bb.get(buf);
    HexDump.dump(buf, 0, System.out, 0);
    */

    byte b = 0;
    Event e = null;
    boolean doneReading = false;

    try {
      while (!doneReading && in.readable()) {
        b = in.readByte();
        switch (m) {
        case START:
          if (b == '<') {
            baos.write(b);
            m = Mode.PRIO;
          } else if(b == '\n'){
            //If the character is \n, it was because the last event was exactly
            //as long  as the maximum size allowed and
            //the only remaining character was the delimiter - '\n', or
            //multiple delimiters were sent in a row.
            //Just ignore it, and move forward, don't change the mode.
            //This is a no-op, just ignore it.
            logger.debug("Delimiter found while in START mode, ignoring..");

          } else {
            isBadEvent = true;
            baos.write(b);
            //Bad event, just dump everything as if it is data.
            m = Mode.DATA;
          }
          break;
        case PRIO:
          baos.write(b);
          if (b == '>') {
            if (prio.length() == 0) {
              isBadEvent = true;
            }
            m = Mode.DATA;
          } else {
            char ch = (char) b;
            prio.append(ch);
            // Priority is max 3 digits per both RFC 3164 and 5424
            // With this check there is basically no danger of
            // boas.size() exceeding this.maxSize before getting to the
            // DATA state where this is actually checked
            if (!Character.isDigit(ch) || prio.length() > 3) {
              isBadEvent = true;
              //If we hit a bad priority, just write as if everything is data.
              m = Mode.DATA;
            }
          }
          break;
        case DATA:
          // TCP syslog entries are separated by '\n'
          if (b == '\n') {
            e = buildEvent();
            doneReading = true;
          } else {
            baos.write(b);
          }
          if(baos.size() == this.maxSize && !doneReading) {
            isIncompleteEvent = true;
            e = buildEvent();
            doneReading = true;
          }
          break;
        }

      }

      // UDP doesn't send a newline, so just use what we received
      if (e == null && isUdp) {
        doneReading = true;
        e = buildEvent();
      }
    } finally {
      // no-op
    }

    return e;
  }

  public Integer getEventSize() {
    return maxSize;
  }

  public void setEventSize(Integer eventSize) {
    this.maxSize = eventSize;
  }

  public void setKeepFields(Set<String> keepFields) {
    this.keepFields = keepFields;
  }
}


